proposals/NNNN-spirv-types-in-clang.md

```diff
@@ -0,0 +1,113 @@
+<!-- {% raw %} -->
+
+# Representing SpirvType in Clang's Type System
+
+- Proposal: [NNNN](NNNN-spirv-types-in-clang.md)
+- Author(s): [Cassandra Beckley](https://github.com/cassiebeckley)
+- Status: **Design In Progress**
+
+## Introduction
+
+We are implementing `SpirvType` and `SpirvOpaqueType` defined in
+[Inline SPIR-V (HLSL proposal 0011)](https://github.com/microsoft/hlsl-specs/blob/main/proposals/0011-inline-spirv.md#types).
+In order to support this feature, we need a way to represent inline SPIR-V types
+in the Clang type system. To facilitate this, I am proposing
+`HLSLInlineSpirvType` as a new `Type` subclass, and `__hlsl_spirv_type` as a new
+builtin template to create such a type.
+
+## Motivation
+
+In DXC, we represented `SpirvType` as a template for a struct. This was simple
+to implement, since in DXC we lower the AST directly to SPIR-V without going
+through LLVM codegen, and could add special cases handling structs that were
+`SpirvType` specializations while lowering types. However, this approach does
+not interact well with Sema and the rest of Clang, resulting in at least several
+bugs in handling template arguments and other areas. In order to take the same
+approach in Clang, we would need to add special case handling while lowering
+struct types, as well as for calculating size and alignment and anywhere else
+structs are interacted with in Clang. Struct types are already processed
+separately from all other types while lowering, and adding special cases for
+`SpirvType` would complicate their handling even more.
+
+Finally, a struct is not a good semantic representation of a `SpirvType`.
+`SpirvType` is fundamentally a new kind of type for Clang that is different from
+all existing types—it represents a low-level target type specified by the user.
```

Comment:
We should mention that It is an opaque type. Representing as a struct does not work because Clang does not treat structs as opaque.

Reply:
Done.